Use AES256 for the default encryption algoritm for TLS session ticket…

Castaglia opened the pull request:
…s, rather

than AES128.  This involves providing more session ticket key data, for both
the cipher and the digest: 80 bytes total = key name(16) + cipher key(32) +
HMAC key(32).

This addresses Issue #514 .

kroeckx commented:
This seems to be failing with:
*** glibc detected *** ./ssltest: free(): invalid pointer: 0x00000000027f609c ***

kroeckx commented on ssl/statem/statem_srvr.c:
```diff
@@ -3022,10 +3022,10 @@ int tls_construct_new_session_ticket(SSL *s)
         if (RAND_bytes(iv, 16) <= 0)
             goto err;
-        if (!EVP_EncryptInit_ex(ctx, EVP_aes_128_cbc(), NULL,
+        if (!EVP_EncryptInit_ex(&ctx, EVP_aes_256_cbc(), NULL,
```
This & at least looks suspicious.

Castaglia replied:
Sorry about that.  Good catch, thanks!

kroeckx replied:
You have the same problem in the other file.

kroeckx commented on ssl/ssl_lib.c:
```diff
@@ -2335,10 +2335,10 @@ SSL_CTX *SSL_CTX_new(const SSL_METHOD *meth)
     ret->max_send_fragment = SSL3_RT_MAX_PLAIN_LENGTH;
 
-    /* Setup RFC4507 ticket keys */
+    /* Setup RFC5507 ticket keys */
```
You probably mean RFC5077

Castaglia replied:
Indeed.  Fixed.

kroeckx commented on ssl/s3_lib.c:
```diff
@@ -3524,19 +3524,19 @@ long ssl3_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
-                memcpy(keys + 16, ctx->tlsext_tick_hmac_key, 16);
-                memcpy(keys + 32, ctx->tlsext_tick_aes_key, 16);
+                memcpy(keys + 16, ctx->tlsext_tick_hmac_key, 32);
+                memcpy(keys + 48, ctx->tlsext_tick_aes_key, 32);
```
Can you replace all those numbers like 16, 32, 48, 80 with something that makes sense, like sizeof(SSL_CTX_get_tlsext_ticket_keys)?

Castaglia replied:
Will do.  As a first pass, I'd wanted to keep in the same style as the existing code and get the approach reviewed, before refining further.

kroeckx commented on ssl/statem/statem_srvr.c:
```diff
@@ -3022,10 +3022,10 @@ int tls_construct_new_session_ticket(SSL *s)
                                 tctx->tlsext_tick_aes_key, iv))
             goto err;
-        if (!HMAC_Init_ex(hctx, tctx->tlsext_tick_hmac_key, 16,
+        if (!HMAC_Init_ex(hctx, tctx->tlsext_tick_hmac_key, 32,
```
That 32 should probably also change to sizeof(tctx->tlsext_tick_hmac_key)

kroeckx commented on ssl/t1_lib.c:
```diff
@@ -3095,15 +3095,17 @@ static int tls_decrypt_ticket(SSL *s, const unsigned char *etick,
         /* Check key name matches */
-        if (memcmp(etick, tctx->tlsext_tick_key_name, 16)) {
+        if (memcmp(etick, tctx->tlsext_tick_key_name,
+                   sizeof(tctx->tlsext_tick_key_name))) {
```
Can you change to that != 0?

Castaglia replied:
Done.

ALSchwalm commented:
Is this still being worked on? I'm interested in seeing this (or any fix for #514) land. I think the problem travis is having at the moment is that it should be `sizeof(ctx->tlsext_tick_aes_key)` on line 3527.

Castaglia commented:
I've updated the PR, and fixed that particular build issue (thanks!); it now appears that the Travis CI builds are failing for something unrelated to my changes.  Hopefully someone on the OpenSSL team could help diagnose those?

kroeckx commented:
On Thu, Mar 17, 2016 at 07:12:19PM -0700, Adam Schwalm wrote:
> Is this still being worked on?

I've already moved it to the internal review some time ago.  I
think I've fixed at least 1 bug in it.

kroeckx commented:
So this got merged.  Maybe you also want to create a merge request for 1.0.2?

Castaglia commented:
@kroeckx done; see PR #1080 .  Thanks!
